### PR TITLE
Use vCenter Event History Collector

### LINKED ISSUE #120: Support event replay (at-least-once semantics)
Currently there is no support for redelivering events in case of cloud events `send` fails after retries (no ACK), the vsphere source crashes/reboots, etc. Supporting `at-least-once` event delivery semantics would be desirable so that we don't skip/lose vsphere events in these scenarios.

A simple approach could be local file checkpointing (with optional strong persistence) to capture the last successfully ACK-ed event. E.g.  in VEBA we use this file structure:

```json
{
  "vCenter": "10.0.0.1",
  "lastEventUUID": "f8d71cb0-fd66-4e3b-aae1-b62d4576bf40",
  "lastEventKey": 26071,
  "lastEventType": "UserLogoutSessionEvent",
  "lastEventKeyTimestamp": "2020-10-13T08:05:36.792Z",
  "createdTimestamp": "2020-10-13T08:05:55.534638Z"
}
```

This file is read after startup and the event stream is started from `lastEventKeyTimestamp` (starting from key/offset is not supported in the public API). A time limit range can be set to have an upper bound for the starting point, e.g. `t.Now()-1h`.

Related: #119

### PR DESCRIPTION
**Pull request by embano1**
With this change the govmomi event manager is replaced with the event history collector which gives more capabilities around filtering and event (stream) replay. Today, only filtering by start time is supported (to implement the event replay feature).

Event history collector uses polling thus backoff logic (fixed, up to 5s) was implemented to avoid aggressive polling. At maximum 100 events will be read per batch from vCenter in one call (hardcoded, will need to see if this is good enough or needs to be configurable for larger environments).

### New features

- Support for event replay via checkpointing (KVStore via ConfigMap)
- Configurable checkpointing behavior (max replay window, checkpoint frequency) via `"VSPHERE_CHECKPOINT_CONFIG"` JSON-encoded string (example below), e.g. environment variable or injected from ConfigMap
- More unit and e2e tests added around event stream behavior (using govmomi simulator)

Checkpoint config example via environment variable:

### Example CheckpointConfig

```bash
# replay from last hour (if available in vCenter), checkpoint every 10s
export VSPHERE_CHECKPOINT_CONFIG={\"maxAge\":\"1h\",\"period\":\"10s\"}
```

> **Note:** Checkpoint config values must be human-readable time values, e.g. `10s`, `5m` or `12h`.

### Example Checkpoint

```bash
{"vCenter":"10.161.153.226","lastEventKey":7433,"lastEventType":"UserLoginSessionEvent","lastEventKeyTimestamp":"2021-02-02T20:06:57.817Z","createdTimestamp":"2021-02-02T20:06:59.286613Z"}
```

### Event Metadata Structure

Here is an example event emitted by this adapter:

```bash
{'attributes': {'specversion': '1.0', 'id': '7456', 'source': '10.161.153.226', 'type': 'com.vmware.vsphere.UserLogoutSessionEvent', 'datacontenttype': 'application/xml', 'time': '2021-02-02T20:17:13.759Z', 'eventclass': 'event'}, 'data': b'<UserLogoutSessionEvent><key>7456</key><chainId>7456</chainId><createdTime>2021-02-02T20:17:13.759Z</createdTime><userName>VSPHERE.LOCAL\\vpxd-extension-43413451-668b-446e-98b1-03bb2d8ce192</userName><fullFormattedMessage>User VSPHERE.LOCAL\\vpxd-extension-43413451-668b-446e-98b1-03bb2d8ce192@127.0.0.1 logged out (login time: Tuesday, 02 February, 2021 07:47:13 PM, number of API invocations: 0, user agent: cl/1.0.0)</fullFormattedMessage><ipAddress>127.0.0.1</ipAddress><userAgent>cl/1.0.0</userAgent><sessionId>52769d8f-e28c-0048-766c-52a119c5d900</sessionId><loginTime>2021-02-02T19:47:13.595101Z</loginTime></UserLogoutSessionEvent>'}
```

Event Schema (CloudEvents v1):

`event.Type`: `com.vmware.vsphere.<EventName>` (e.g. `com.vmware.vsphere.VmPoweredOffEvent`)
`event.ID`: `vCenter.Event.Key` (e.g. `1022`)
`event.Timestamp`: `vCenter.Event.CreatedTime`
`event.Subject`: (unused for now)
`event.Source`: vCenter URL this source adapter is linked to
`event.EventClass` (Extended CE attribute): e.g. `Event`, `EventEx` or `ExtendedEvent`
`event.DataContentType`: `application/xml` (currently fixed)

Closes: #120 #119 #118

Signed-off-by: Michael Gasch <mgasch@vmware.com>

### CODE REVIEW

File: pkg/vsphere/adapter.go [hunk 1]
```diff
@@ -80,41 +97,192 @@ func (a *vAdapter) Start(ctx context.Context) error {
...
+
+	cpTicker := time.NewTicker(a.CpConfig.Period)
+	defer cpTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		// 	checkpoints
```
Comment (vaikas) [inline pkg/vsphere/adapter.go:158]: nit funny spacing for the comment here and below

File: pkg/vsphere/adapter.go [hunk 2]
```diff
@@ -80,41 +97,192 @@ func (a *vAdapter) Start(ctx context.Context) error {
...
+	begin := vcTime
+	logger := logging.FromContext(ctx)
+
+	cpTime := cp.LastEventKeyTimestamp
+	if !cpTime.IsZero() {
+		// valid checkpoint
+		logger.Info("found existing checkpoint")
+		maxTime := begin.Add(maxAge * -1)
+		if maxTime.Unix() > cpTime.Unix() {
+			logger.Warnw("last event timestamp in checkpoint is older than configured maximum", "maxHistory", maxAge.String(), "checkpointTimestamp", cp.LastEventKeyTimestamp.String())
```
Comment (vaikas) [inline pkg/vsphere/adapter.go:275]: nit, might want to include in the log message that this means some events can be lost?

**Review by vaikas — COMMENTED** [review]
Overall looking great. Few nits and suggestions.

**Comment by vaikas** [comment]
Looks like you need to run ./hack/update-codegen.sh

File: pkg/vsphere/checkpoint.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package vsphere
+
```
[the remark this answers is not in the document]
Comment (embano1) [reply pkg/vsphere/checkpoint.go:2]: Ran hack scripts but they did not inject, manually added them...

File: pkg/vsphere/adapter.go (hunk 2 above)
Comment (embano1) [reply pkg/vsphere/adapter.go:275]: Good point, explicit now:

`logger.Warnw("potential data loss: last event timestamp in checkpoint is older than configured maximum",`

**Comment by embano1** [comment]
Thx for the quick review! Yeah, had to fix license issue due to old bits on my machine. Then `code-gen` was fine.
Made all the required changes, but before merging let's please discuss the open topics in the PR message above.

**Comment by embano1** [comment]
All WIP items clarified and documented in PR. Ready to merge @vaikas

**Review by vaikas — APPROVED** [review]

**Comment by vaikas** [comment]
Thanks for doing this, in a followup, we should also update README with the examples you give in the description for this PR.

**Comment by embano1** [comment]
Good point, let me create follow-up issue!

**Pull request merged**